### PR TITLE
Implements the Authentication Configuration Checks

### PR DESCRIPTION
**Pull request by mereghost**
### Related Work Package: [OP#62100](https://community.openproject.org/work_packages/62100)

Implements the OIDC connection tests and some elementary OAuth2 tests.

### CODE REVIEW

File: modules/storages/app/common/storages/peripherals/connection_validators/nextcloud/authentication_validator.rb [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+              @storage.authenticate_via_idp? ? validate_sso : validate_oauth
+            end
+
+            @results
+          end
+
+          private
+
+          def validate_oauth
+            @results = {
```
Comment (NobodysNightmare) [inline modules/storages/app/common/storages/peripherals/connection_validators/nextcloud/authentication_validator.rb]: :red_circle:  The usage of the `@results` instance variables seems to be all over the place to me. `BaseValidator` relies on the fact that it is defined when using `pass_check`, `warn_check` and `fail_check`. However, `BaseValidator didn't make sure that this instance variable exists.

`call` (in a subclass of `BaseValidator`) returns the instance variable, but it didn't fill in anything itself, relying on the variable being set by both methods called from it.

Setting the initial value happens in two different places for this subclass and in one method in another subclass.

That was the problem statement from my side, I think we should make sure that access to `@results` is better controlled.

---

Given what I currently see, my preferred solution would probably be to never use `@results` outside of  `BaseValidator`. We could add a single method to `BaseValidator` (name is a proposal):

```ruby
def register_check(name) # or prepare_check or register_result or ...
  name = name.to_sym
  @results[name] = CheckResult.skipped(name)
end

# the initializer would initialize an empty hash
```

Then subclasses would just call this method, e.g.

```ruby
register_check :non_provisioned_user
register_check :non_oidc_provisioned_user
register_check :token_usability

# if you are hungry for a nice DSL you could also allow many names
register_checks :a, :b, :c
```

As a last "problem" we'd need to figure out what to do with `call`.

Option A: `call` does not return results, you validate differently:

```ruby
validator.call
validator.results
```

Option B: `call` just returns `results`, which are preferably made accessible through an attribute reader.

Option C: `call` is actually implemented in `BaseValidator` and subclasses implement validations in a different method. This is my favorite approach. Example follows:

```ruby
class BaseValidator
  def call
    @results = {} #resetting results in built-into call
    catch :interrupted do
      validate
    end
    @results
  end

  def validate; raise SubclassResponsibility; end
end

class MyValidator < BaseValidator
  def validate
    register_check :foo
    fail_check :foo
  end
end
```

I just realize that Option C solves a problem I overlooked so far: Handling the throwable `interrupted` was **also** all over the place: The base class threw it and the subclasses needed to know to catch it. In Option C, only `BaseValidator is concerned with it and subclasses can focus on validation logic with a nice DSL.

File: modules/storages/app/common/storages/peripherals/connection_validators/nextcloud/authentication_validator.rb [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+            oauth_token
+            user_bound_request
+          end
+
+          def oauth_token
+            if OAuthClientToken.where(user: @user, oauth_client: @storage.oauth_client).any?
+              pass_check(:existing_token)
+            else
+              warn_check(:existing_token, message(:oauth_token_missing))
+              throw :interrupted
```
Comment (NobodysNightmare) [inline modules/storages/app/common/storages/peripherals/connection_validators/nextcloud/authentication_validator.rb]: :yellow_circle: In accordance with my earlier stated opinion that we shouldn't throw (pun intended) `interrupted` around everywhere, what do you think about exposing this as an API, e.g. through `warn_check(:existing_token, message, critical: true)` or `critical_warn_check(...)`?

It would be great if subclasses wouldn't need to know the intricates of which symbols are thrown in a validator. Similarly to how I don't need to know how Sinatra's `halt` is implemented (probably also achievable by throwing just the right thing).

File: modules/storages/app/common/storages/peripherals/connection_validators/nextcloud/authentication_validator.rb [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+            end
+          end
+
+          def user_bound_request
+            Registry["nextcloud.queries.files"]
+              .call(storage: @storage, auth_strategy:, folder: ParentFolder.new("/")).on_failure do
+              fail_check(__method__, message("oauth_request_#{it.result}"))
```
Comment (NobodysNightmare) [inline modules/storages/app/common/storages/peripherals/connection_validators/nextcloud/authentication_validator.rb]: :green_circle: A bit much meta-programming for my taste, since it doesn't seem to serve any other purpose than to avoid repeating the method name, which _happens to coincide_ with the name of a check we registered earlier.

Possible pitfalls that I see:
* searching for `:user_bound_request` does not show me this place (even though I saw that we registered this symbol as a test name earlier)
* Renaming the method now has the side effect of us not updating the correct check result anymore

IMO it's just not worth the meta.

File: modules/storages/app/common/storages/peripherals/connection_validators/nextcloud/authentication_validator.rb [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+
+            oauth_token
+            user_bound_request
+          end
+
+          def oauth_token
+            if OAuthClientToken.where(user: @user, oauth_client: @storage.oauth_client).any?
+              pass_check(:existing_token)
+            else
+              warn_check(:existing_token, message(:oauth_token_missing))
```
Comment (Kharonus) [inline modules/storages/app/common/storages/peripherals/connection_validators/nextcloud/authentication_validator.rb]: ❓ why is it `warn_check` and not a `fail_check`?

File: modules/storages/app/common/storages/peripherals/connection_validators/nextcloud/authentication_validator.rb [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+                         in { code: /token_refresh/ }
+                           :oidc_cant_refresh_token
+                         in { code: :no_token_for_audience }
+                           :oidc_cant_acquire_token
+                         else
+                           :unknown_error
+                         end
+
+            fail_check(__method__, message(error_code))
+            throw :interrupted
```
Comment (Kharonus) [inline modules/storages/app/common/storages/peripherals/connection_validators/nextcloud/authentication_validator.rb]: 🟡 no need to throw, if `fail_check` was called

File: modules/storages/app/common/storages/peripherals/connection_validators/nextcloud/authentication_validator.rb (hunk 4 above)
Comment (Kharonus) [reply modules/storages/app/common/storages/peripherals/connection_validators/nextcloud/authentication_validator.rb]: same in `non_provisioned_user` and `non_oidc_provisioned_user`

File: modules/storages/app/common/storages/peripherals/connection_validators/nextcloud/authentication_validator.rb (hunk 5 above)
Comment (NobodysNightmare) [inline modules/storages/app/common/storages/peripherals/connection_validators/nextcloud/authentication_validator.rb]: :yellow_circle: Unreachable code. `fail_check` already throws.

**Review by NobodysNightmare — CHANGES_REQUESTED** [review]
In addition to the :red_circle: that I cleverly posted immediately, I added all other traffic light colors to round it off.

In general I like what those validators might look like soonish.

File: modules/storages/app/common/storages/peripherals/connection_validators/nextcloud/authentication_validator.rb (hunk 4 above)
Comment (NobodysNightmare) [reply modules/storages/app/common/storages/peripherals/connection_validators/nextcloud/authentication_validator.rb]: I guess because the check doesn't quite mean that stuff is entirely broken, but rather "I can't check further, that makes me sad". So we need to abort, but we can't say for sure that it's broken.

File: modules/storages/app/common/storages/peripherals/connection_validators/nextcloud/authentication_validator.rb (hunk 2 above)
Comment (NobodysNightmare) [reply modules/storages/app/common/storages/peripherals/connection_validators/nextcloud/authentication_validator.rb]: More naming ideas after writing in Eric's thread:

* `warn_check(..., abort: true`)
* `abort_with_warning(...)`

File: modules/storages/app/common/storages/peripherals/connection_validators/nextcloud/authentication_validator.rb (hunk 4 above)
Comment (mereghost) [reply modules/storages/app/common/storages/peripherals/connection_validators/nextcloud/authentication_validator.rb]: Yep, it is not a failure as I can't simply get a token without going through the entire OAuth2 process.

This makes the context slightly different. =/

File: modules/storages/app/common/storages/peripherals/connection_validators/nextcloud/authentication_validator.rb (hunk 5 above)
Comment (mereghost) [reply modules/storages/app/common/storages/peripherals/connection_validators/nextcloud/authentication_validator.rb]: Oh! I missed that.

File: modules/storages/app/common/storages/peripherals/connection_validators/nextcloud/authentication_validator.rb (hunk 3 above)
Comment (mereghost) [reply modules/storages/app/common/storages/peripherals/connection_validators/nextcloud/authentication_validator.rb]: <del>I'm keeping it as so, as</del> I feel that I'll end up renaming all methods and keys once I gather all the results in one place.

Specially as the names represented negative outcomes as we where chaining calls and now we have proper checks with 2 relevant different outcomes.

File: modules/storages/app/common/storages/peripherals/connection_validators/nextcloud/authentication_validator.rb (hunk 1 above)
Comment (mereghost) [reply modules/storages/app/common/storages/peripherals/connection_validators/nextcloud/authentication_validator.rb]: I fully agree with you on this. 🤣 

I'm playing with how to make it nicer overall, but I had to have some stuff in place so I can see where to take it and what would make sense.

My original idea was identical to yours: have a way to register the checks and never have to worry about the results. After a couple of tests, I concluded that I needed the actual implementations before I could extract this behaviour.

`pass_check` (etc) was born as the start of this and I'm slowly cleaning up the interface, as `BaseValidator` was just introduced to start to gobble those details.

I was already moving towards what you propose for the AMPF part of the validations.

File: modules/storages/app/common/storages/peripherals/connection_validators/nextcloud/authentication_validator.rb (hunk 3 above)
Comment (mereghost) [reply modules/storages/app/common/storages/peripherals/connection_validators/nextcloud/authentication_validator.rb]: Removed the calls to `__method__`.

File: modules/storages/app/common/storages/peripherals/connection_validators/nextcloud/authentication_validator.rb (hunk 2 above)
Comment (mereghost) [reply modules/storages/app/common/storages/peripherals/connection_validators/nextcloud/authentication_validator.rb]: I added the argument `halt_validation` and didn't add a specialized method for now.

File: modules/storages/app/common/storages/peripherals/connection_validators/nextcloud/base_validator.rb [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+            @results
+          end
+
+          private
+
+          def validate = raise Errors::SubclassResponsibility
+
+          def register_checks(*keys)
+            @results = keys.to_h { [it, CheckResult.skipped(it)] }
```
Comment (NobodysNightmare) [inline modules/storages/app/common/storages/peripherals/connection_validators/nextcloud/base_validator.rb:53]: :green_circle: The current implementation of this implies that I can only call `register_checks` once, never twice.
 
I think that's enough, but it limits the ways in which I can write my validator a bit. E.g. this layout would not work:

```ruby
def validate
  register_checks(:common_check)
  if rand < 0.5
    register_checks(*a_checks)
    validate_a
  else
    register_checks(*b_checks)
    validate_b
  end
end
```

The other issue is that if I for whatever reason don't add any tests, we'd now return `nil` instead of returning `{}`. As indicated by the green traffic light, I think both of these problems are absolutely minor, but I wanted to explicitly write them down at least.

**Review by NobodysNightmare — APPROVED** [review]
